### ISSUE TITLE
shadow casting issues

### ISSUE THREAD

**Issue by Hogarth-MMD** [powroupi/blender_mmd_tools#83]
mmd_tools has a weird, mysterious issue with spot lamp buffer shadows. I tested this very carefully. I imported the exact same pmx model (Byakuren Hijiri by Pluto) into the same scene with both mmd_tools and with Pymeshio. The 2 models standing next to each other in the same scene. The model imported with mmd_tools does not cast any spot lamp buffer shadow. The model imported with pymeshio casts a spot lamp buffer shadow, no problem.

**Comment by Hogarth-MMD** [comment]
Both models do cast a shadow when rendered, but the model imported with mmd_tools does not cast any preview shadow in the 3d viewport.

**Comment by nathanvasil** [comment]
It's not impossible that MMD Tools is handling this correctly and Pymeshio is handling it incorrectly.  I can't download the model because it's passworded.  What are its .pmx self-shadow settings?  MMD Tools reads these as it should, but Pymeshio may not, and just assign shadows to all materials regardless of settings.

In Blender, check the material settings on the imported model.  First, look under MMD Material.  Is Self Shadow Map checked?  Look the Shadow section of the Materials tab.  Is Cast checked?  Is Cast Buffer Shadows checked?

**Comment by Hogarth-MMD** [comment]
I am not aware of Pluto models being password-protected. Download links are here in comments:
https://hogarth-mmd.deviantart.com/journal/Where-to-download-MMD-models-631052349

**Comment by Hogarth-MMD** [comment]
Yes Cast Shadow and Cast Buffer Shadow are both checked, which is why I said that it is mysterious.

**Comment by nathanvasil** [comment]
Still didn't download that model, because the link is a list that links to another list that's all in Japanese, so I can't easily locate the model.

Instead I loaded the .blend you provided and imported a model.

So long as the object contains any materials for which "cast" is unchecked, the entire object will not cast buffer shadows in material view.  It doesn't matter whether any faces are assigned to those materials or not.

"Cast buffer shadows" doesn't appear to matter, at least in this view.

This is probably an issue where Pymeshio is failing to read the .pmx value and so avoiding the issue, but the issue is with Blender, not MMD tools (or possibly with that specific blend file, I'm not sure.)  If you make a cube in that scene, create two materials for it, and disable "cast" on one of those materials, it will stop casting buffer shadows in the material view.  Again, doesn't even matter if any faces are assigned.

If you import a model and separate by materials, individual objects will end up each with a single material and the model will cast shadows appropriately in material view.

If this is important to fix, you could consider importing MMD models separated by materials to start with, possibly with an automatically generated group to make it easier to handle the entire mesh separate from the armature.  You could also consider submitting a bug report for Blender, although this may be by design, as a means to optimize preview display.

All mysteries are mysteries exactly until they're not, at which point they are obvious.  Properly troubleshooting anything requires making fewer assumptions than what we're generally accustomed to.

Edit: Checked Pymeshio.  Sure enough, Pymeshio is checking "cast" for all materials without regard to the self-shadow .pmx property.

**Comment by powroupi** [comment]
Yes, as @nathanvasil said, it is a preview issue of Blender 3d viewport. And as I mentioned [here](https://github.com/powroupi/blender_mmd_tools/issues/57#issuecomment-346921375), separate by materials will lead to an issue of transparency draw order in 3d viewport preivew. It's easy to change the material draw order of a object but it requires using some tricks to change object draw order.

And there are some limitations of blender 3d viewport realtime preview. Some interesting tutorials: [hacking-the-viewport](https://www.blendernation.com/2014/12/18/blender-2-7-hacking-the-viewport/). :smile:

**Comment by Hogarth-MMD** [comment]
"So long as the object contains any materials for which "Cast" is unchecked, the entire object will not cast buffer shadows in material view."

This is the correct answer. Thank you @nathanvasil for taking the time to deduce this correct answer. @powroupi did not know the correct solution to this issue. So, this is an mmd_tools bug, is it not? Every material should have "Cast" enabled when an MMD model is imported into Blender?

**Comment by Hogarth-MMD** [comment]
Here a different, but related: issue:
The model AB Chan 7, by Kakomiki has an alternate body material which is named '\\body'. This material is invisible by default, but it has "Cast" enabled, which causes a render of this model to look like her body is covered with dirt. If "Cast" is disabled, this issue of render covered with dirt is solved.

**Comment by Hogarth-MMD** [comment]
So, this is an mmd_tools bug, is it not? Every material WHICH IS NOT INVISIBLE should have "Cast" enabled when an MMD model is imported into Blender?

**Comment by Hogarth-MMD** [comment]
So, if an MMD model has even one invisible material, then apparently we have no solution to the buffer shadows issue for that model(?).

**Comment by Hogarth-MMD** [comment]
I just did this simple experiment in Blender 2.78, not using mmd_tools, just creating a cube and a plane:
The cube will cast a preview render of a shadow on the plane, in both texture view and material view, using either a spot lamp or a sun lamp. If the cube has even one material without Cast enabled, no preview shadow. Tutorials are saying that spot lamp with buffer shadows are needed to preview render a shadow, but I am seeing a preview shadow render with sun lamp.

**Comment by nathanvasil** [comment]
> So, this is an mmd_tools bug, is it not? Every material WHICH IS NOT INVISIBLE should have "Cast" enabled when an MMD model is imported into Blender?

No, I wouldn't say that this is an mmd_tools bug.  I would say that it is a Blender bug.  I believe mmd_tools is handling this in the best way that it can without making basic Blender bug fixes, which should be handled by the people working on Blender.

Note that if you enable "cast" by default, you're going to have problems as well, and they're going to show up more frequently than if you determine material settings from .pmx material settings.  Any model with eye highlights is going to look terrible, as their eyes suddenly get shadowed by these big planes.  And there are a lot of models with eye highlights.  Eye highlights aren't the only place where it's important to prevent shadowing.

> his material is invisible by default, but it has "Cast" enabled

This is something that mmd tools could potentially handle.  IIRC, materials with opacity 0 don't get shadowed by the  MMD default shader, although this isn't handled in the HLSL, probably in the C or whatever instead.  Would it be possible to disable cast on material opacity 0 materials, and enable them with whatever driver is used for the material morph that changes their opacity?

>So, if an MMD model has even one invisible material, then apparently we have no solution to the buffer shadows issue for that model(?).

The solution is to separate the model by materials.  You can do that manually, or you could make it the default (I wouldn't mind, that's the first thing I always do.)  Currently, it doesn't look like there's any way for Blender to properly handle a single object with multiple materials with different Cast characteristics.

**Comment by powroupi** [comment]
>Would it be possible to disable cast on material opacity 0 materials, and enable them with whatever driver is used for the material morph that changes their opacity?

Yes, this is possible I think, I'll try to fix it later. :smile:

**Comment by powroupi** [comment]
Okay, it is fixed now. The "cast" is disabled on material opacity 0 materials, but looks like only Blender 2.78+ respect "cast" in 3d viewport preview (render view is correct though), order versions will still have this issue. :sweat_smile:

**Comment by nathanvasil** [comment]
> The "cast" is disabled on material opacity 0 materials

Tested to make sure it's handling export properly.  Looks great, preserving shadow settings on opacity 0 mats through an import/export.

Also tested to see if "receive" was something that should be treated similarly.  Doesn't appear to matter, received shadows get the alpha 0 and so don't show up in the render.  Tested with 2.78c Blender Internal f12 render.

**Comment by powroupi** [comment]
Thanks for the test, @nathanvasil :smile: 

Here is the current table for shadow settings:

MMD Shadow | Blender Shadow
------------------ | ---------------------
Self Shadow | Recieve & Recieve Transparent
Self Shadow Map | Cast & Cast Buffer Shadows (Disabled if alpha < 0.001)
Ground Shadow | Ignored

**Comment by Hogarth-MMD** [comment]
Thank you for the bug fix and for all your efforts, @powroupi . I wish I could think of a better solution that guarantees easy, user-friendly preview shadows like we have in MMD.

**Comment by Hogarth-MMD** [comment]
How about this? A material does not have Cast enabled, which prevents a preview shadow. In the Material panel, where it says Link Data Object, change from Data to Object. Now the preview shadow appears. Does this work, or am I full of baloney with this solution? :smile:

**Comment by Hogarth-MMD** [comment]
Hmm, Data to Object causes the material to lose its transparency and appear all white, so I guess that doesn't work.

**Comment by Hogarth-MMD** [comment]
Split by materials is irreversible, is it not? To be a good solution, it should  preferably be temporary and reversed with one click.

**Comment by nathanvasil** [comment]
>Split by materials is irreversible, is it not? To be a good solution, it should preferably be temporary and reversed with one click.

IIRC, "join meshes" button, with any component of a separated MMD model selected, will reassemble all component objects into a single object.

I'm not sure what benefit exists in having all materials as a single object.  But then, I use Blender and MMD Tools for modelling, texturing, etc, not for animation or rendering.  I may not be aware of all of the issues.

I can think of one situation that might be a problem, which is that UV morphs are lost on separation by materials.  I think.  I'd be surprised if this is something for which no imaginable solution exists, but I don't know enough to be sure.

**Comment by Hogarth-MMD** [comment]
Splitting to 2 objects, casting and non-casting would be better.

**Comment by Hogarth-MMD** [comment]
I think that maybe I have found a solution to this problem. Enable "Cast" on every material. Now the object will cast a preview shadow, but we have the problem of transparent materials causing a render covered with dirt. But we can solve this problem of render covered with dirt by disabling "Traceable" on these materials.

**Comment by nathanvasil** [comment]
Hmm.  In my testing, an object with "cast" enabled but "traceable" disabled still casts shadows in the material view.  It doesn't in rendered view, butthen, rendered view was never the problem.  Not that this skirt shouldn't be casting a shadow.  This will be a problem for the material view shadows especially in situations that use overlaid planes with alpha channels that aren't supposed to cast shadows, the kinds of meshes used for things like eye highlights and decals.  (I use these to get smooth neck, lip transitions that allow me to have a low specular power face, a moderate specular power body, and high specular power lips.)
![shadows](https://user-images.githubusercontent.com/19930899/33445264-41e54d8e-d5b1-11e7-8a6d-ac71454ae856.png)


BTW, I just want to say thanks for sharing all your thoughts on this as you go.  By following along, I'm learning a lot.

**Comment by Hogarth-MMD** [comment]
I just found another issue with transparent materials. Under Transparency there is an Alpha slider and a Specular slider. If the Specular slider is not set to 0.0, the transparent material will still render specular highlights. These "invisible" materials can have specular highlights in addition to casting shadows.

**Comment by Hogarth-MMD** [comment]
In the image posted by @nathanvasil we can see that this "invisible" material is still being displayed as a wire-frame mesh.

**Comment by nathanvasil** [comment]
That's just because I've selected it, so that you could see it.

**Comment by powroupi** [comment]
Another way is to disable render (icon) of invisible object in [Outliner](https://docs.blender.org/manual/en/dev/editors/outliner.html). :)

**Comment by Hogarth-MMD** [comment]
It is not possible to disable the rendering of an individual material in Outliner, though.

**Comment by Hogarth-MMD** [comment]
In the Properties panel at the right side of the 3D view, under Display, there is an option "Only Render". That gets rid of the wire-frame preview of the invisible material.
